src/main/scala/org/kframework/minikore/interfaces/MiniKoreInterface.scala

```diff
@@ -0,0 +1,303 @@
+package org.kframework.minikore.interfaces
+
+object TreeInterface {
+
+  import PatternInterface._
+
+  sealed trait AST
+
+
+  sealed trait Node extends AST with Product {
+    def args: Seq[Pattern]
+
+    def build(children: Seq[Pattern]): Pattern
+  }
+
+  object Node {
+    def unapply(arg: Node): Option[Seq[Pattern]] = Some(arg.args)
+  }
+
+
+  sealed trait Leaf[C] extends AST with Product {
+    def contents: C
+
+    def build(contents: C): Pattern
+  }
+
+  object Leaf {
+    def unapply[C](arg: Leaf[C]): Option[C] = Some(arg.contents)
+  }
+
+
+  sealed trait Leaf2[CC1, CC2] extends Leaf[Product2[CC1, CC2]] with Product2[CC1, CC2] {
+    override def contents: Product2[CC1, CC2] = (_1, _2)
+
+    def apply(_1: CC1, _2: CC2): Pattern
```

Comment:
@cos - I think the apply is most commonly used in the companion object. Given my limited understanding of scala, I don't see why it'd be the convention. From what I've seen, it seems that given an object x, x(y) desugars to x.apply(y). In the case of a class, having a companion object with the same name and apply method is useful, as it avoids having to create an object with the new keyword. However, if I have the right idea, it doesn't matter where apply actually is. The only thing that matters is x(y) desugars to x.apply(y), which works well in our case. If the scala community is enforcing the convention that the apply must be in the companion object, then isn't it limiting it's usage?